### PR TITLE
Add Length, Slice and Invert methods to Delta

### PR DESCRIPTION
**Pull request by xxuejie**
This PR also contains a small bug fix uncovered while implementing Invert()

### CODE REVIEW

**Comment by fmpwizard** [comment]
thanks for the Pr @xxuejie  !!

I see that the `Invert()` method was added to quiljs on https://github.com/quilljs/delta/blob/master/CHANGELOG.md#v420

Do you happen to have a link for discussions on why it was added?

**Comment by xxuejie** [comment]
I don't have a link for that, bit if you think about it, Invert is super helpful when implementing undos: given a base document and a change delta, all you need to do is calculate the invert and keep it, and when you want to do undo later, you can just apply the invert delta on the document, which will perform the undo operation.

I do have a use case for implementing undo in my project , that's why I added this : P

**Comment by xxuejie** [comment]
BTW I did a quick search and found this original issue: https://github.com/quilljs/delta/issues/28

**Comment by fmpwizard** [comment]
awesome! I'll take a better look at the code tomorrow and most likely merge it after that, thanks for the links!

File: delta/op.go [hunk 1]
```diff
@@ -98,6 +100,28 @@ func AttrTransform(a, b map[string]interface{}, priority bool) map[string]interf
...
+	}
+	for k, v := range attr {
+		v2, exists := base[k]
+		if (!exists) && v2 != v {
+			ret[k] = nil
+		}
+	}
+	if len(ret) == 0 {
+		return nil
+	} else {
```
Comment (fmpwizard) [inline delta/op.go]: you can drop the `else` branch and  have `return ret` be the last line

File: delta/delta.go [hunk 2]
```diff
@@ -31,6 +31,17 @@ func (o *Op) IsNil() bool {
 		o.Retain == nil
 }
 
+func (o *Op) Length() int {
```
Comment (fmpwizard) [inline delta/delta.go]: Need exported func comment

File: delta/delta.go [hunk 3]
```diff
@@ -294,3 +305,56 @@ func (d *Delta) Transform(other Delta, priority bool) *Delta {
 
 	return delta.Chop()
 }
+
+func (d *Delta) Length() int {
```
Comment (fmpwizard) [inline delta/delta.go]: Need exported func comment

File: delta/delta.go [hunk 4]
```diff
@@ -294,3 +305,56 @@ func (d *Delta) Transform(other Delta, priority bool) *Delta {
...
+	length := 0
+	for _, op := range d.Ops {
+		length += op.Length()
+	}
+	return length
+}
+
+func (d *Delta) Slice(start int, end int) *Delta {
```
Comment (fmpwizard) [inline delta/delta.go]: Need exported func comment

File: delta/delta.go [hunk 5]
```diff
@@ -294,3 +305,56 @@ func (d *Delta) Transform(other Delta, priority bool) *Delta {
...
+		} else {
+			nextOp = iter.Next(end - index)
+			delta.Push(nextOp)
+		}
+		index += nextOp.Length()
+	}
+	return delta
+}
+
+func (d *Delta) Invert(base *Delta) *Delta {
```
Comment (fmpwizard) [inline delta/delta.go]: Need exported func comment

File: delta/op.go [hunk 6]
```diff
@@ -98,6 +100,28 @@ func AttrTransform(a, b map[string]interface{}, priority bool) map[string]interf
 	return nil
 }
 
+// Invert an attribute map, used in Delta.Invert()
+func AttrInvert(attr map[string]interface{}, base map[string]interface{}) map[string]interface{} {
```
Comment (fmpwizard) [inline delta/op.go]: code comment should start with `// AttrInvert `

**Comment by xxuejie** [comment]
@fmpwizard Thanks! All comments are tackled now.

**Comment by fmpwizard** [comment]
thanks @xxuejie 

I also added a tag v0.0.5 that includes these changes, and also note that last night I also merged 

https://github.com/fmpwizard/go-quilljs-delta/pull/10

**Pull request merged**